Documentation/contributing/development/introducing_new_crds.rst

```diff
@@ -0,0 +1,400 @@
+
+   pkg/k8s/apis/cilium.io/v2/
+   pkg/k8s/apis/cilium.io/v2alpha1/
+
+The portion of the directory after ``apis/`` makes up the CRD’s
+``Group`` and ``Version``. See
+`KubeBuilder-GVK <https://book.kubebuilder.io/cronjob-tutorial/gvks.html>`__
+
+You can begin defining your ``CRD`` structure, making any subtypes you
+like to adequately define your data model and using ``marks`` to control
+the CRD generation process.
+
+Here is an brief example, omitting any further definitions of sub-types
+to express the CRD data model. \`\`
+
+.. code:: go
+
+   // +genclient
+   // +genclient:nonNamespaced
+   // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+   // +kubebuilder:resource:categories={cilium,ciliumbgp},singular="ciliumbgppeeringpolicy",path="ciliumbgppeeringpolicies",scope="Cluster",shortName={bgpp}
+   // +kubebuilder:printcolumn:JSONPath=".metadata.creationTimestamp",name="Age",type=date
+   // +kubebuilder:storageversion
+
+   // CiliumBGPPeeringPolicy is a Kubernetes third-party resource for instructing
+   // Cilium's BGP control plane to create peers.
+   type CiliumBGPPeeringPolicy struct {
+       // +k8s:openapi-gen=false
+       // +deepequal-gen=false
+       metav1.TypeMeta `json:",inline"`
+       // +k8s:openapi-gen=false
+       // +deepequal-gen=false
+       metav1.ObjectMeta `json:"metadata"`
+
+       // Spec is a human readable description of a BGP peering policy
+       //
+       // +kubebuilder:validation:Required
+       Spec CiliumBGPPeeringPolicySpec `json:"spec,omitempty"`
+   }
+
+Integrating CRDs into Cilium
+----------------------------
+
+Once you’ve coded your CRD data model you can use Cilium’s ``Make``
+infrastructure to generate and integrate your CRD into Cilium.
+
+There are several make targets and a script which revolve around
+generating CRD and associated code gen (client, informers, DeepCopy
+impls, DeepEqual impls, etc…)
+
+Each of the next sections also detail the steps you should take to
+integrate your CRD into Cilium.
+
+Generating CRD Yaml
+~~~~~~~~~~~~~~~~~~~
+
+To simply generate the CRDs and copy them into the correct location you
+must perform two tasks \* Update the make file to copy your generated
+CRD from a ``tmp`` directory to the correct location in Cilium
+repository. Edit the following
+`location <https://github.com/cilium/cilium/blob/89ca3eddf3dae9ac5fc6c343a2cd26cf3aa405fa/Makefile#L303-L311>`__
+\* Run ``make manifests``
+
+This will generate your Golang structs into CRD manifests and copy them
+to ``./pkg/k8s/apis/cilium.io/client/crds/`` into the appropriate
+``Version`` directory.
+
+You can inspect your generated ``CRDs`` to confirm they look OK.
+
+Additionally ``./contrib/scripts/check-k8s-code-gen.sh`` is a script
+which will generate both generate the CRD manifest along with generating
+the necessary K8s API changes to use your CRDs via K8s client in Cilium
+source code.
+
+Generating Client Code
+~~~~~~~~~~~~~~~~~~~~~~
+
+``make generate-k8s-api``
```

Comment:
Please format into a code block, `.. code-block:: shell-session`. This looks clearer, and allows for easy copy-paste.